projects/python-client/src/datapane/client/api/ipython_utils.py

```diff
@@ -19,10 +25,219 @@ def block_to_iframe(block: BaseElement) -> str:
+
+    return notebook_json
+
+
+@capture_event("Get Colab Notebook JSON")
+def get_colab_notebook_json() -> dict:
+    """Get the JSON for the current Colab notebook
+
+    Returns:
+        Notebook JSON
+    """
+    import ipynbname
+    from google.colab import auth
+    from googleapiclient.discovery import build
+    from googleapiclient.http import MediaIoBaseDownload
+
+    # Get the notebook's Google Drive file_id
+    file_id = ipynbname.name().replace("fileId=", "")
+
+    try:
+        auth.authenticate_user()
+    except Exception as e:
+        raise DPError(
+            "Google Drive authentication failed. Please allow this notebook to access your Google Drive."
+        ) from e
+
+    drive_service = build("drive", "v3")
+
+    request = drive_service.files().get_media(fileId=file_id)
+    downloaded = io.BytesIO()
+    downloader = MediaIoBaseDownload(downloaded, request)
+    done = False
+    while done is False:
+        # _ is a placeholder for a progress object that we ignore.
+        # (Our file is small, so we skip reporting progress.)
+        _, done = downloader.next_chunk()
+
+    downloaded.seek(0)
+    notebook_json = json.loads(downloaded.read().decode("utf-8"))
+
+    return notebook_json
+
+
+def get_notebook_json() -> dict:
+    """Get the JSON for the current Jupyter, Colab, or VSCode notebook
+
+    Returns:
+        Notebook JSON
+    """
+    if "COLAB_GPU" in os.environ:
+        notebook_json = get_colab_notebook_json()
+    elif "VSCODE_PID" in os.environ:
+        notebook_json = get_vscode_notebook_json()
+    else:
+        notebook_json = get_jupyter_notebook_json()
+
+    return notebook_json
+
+
+def markdown_cell_to_block(cell: dict) -> Text:
+    """Convert a IPython notebook cell to a Datapane Text Block
+
+    Args:
+        cell: IPython notebook cell dict
+
+    Returns:
+        Datapane Block
+    """
+    from .report.blocks import Text
+
+    block = Text("".join(cell["source"]))
+
+    return block
+
+
+def input_cell_to_block(cell: dict) -> Code:
+    """Convert a IPython notebook cell to a Datapane Code Block
+
+    Args:
+        cell: IPython notebook cell dict
+
+    Returns:
+        Datapane Block
+    """
+    from .report.blocks import Code
+
+    block = Code("".join(cell["source"]))
+
+    return block
+
+
+def output_cell_to_block(cell: dict, ipython_output_cache: dict) -> typing.Any[BaseElement, None]:
+    """Convert a IPython notebook output cell to a Datapane Block
+
+    Args:
+        cell: IPython notebook cell dict
+
+    Returns:
+        Datapane Block
+    """
+    from .report.blocks import wrap_block
+
+    # Get the output object from the IPython output cache
+    cell_output_object = ipython_output_cache.get(cell["execution_count"], None)
+
+    # If there's corresponding output object, skip
+    if cell_output_object is None:
+        return None
+
+    try:
```

Comment:
```python
with suppress(Exception):
    return wrap_block(cell_output_object)
```

also, should we be catching `Exception` here?

Reply:
I've updated this because I can see we're using `suppress` elsewhere in the codebase. However, I did find [this](https://github.com/django/django/commit/6e4c6281dbb7ee12bcdc22620894edb4e9cf623f), where Django reverted back to `try`/`except` because it performed better!